TaskTest."nmap6 must run Tasks in parallel" is broken

[This test](https://github.com/scalaz/scalaz/blob/a30d3fa8a9a681d99ab1fd42c13edd4ee0e4c19d/tests/src/test/scala/scalaz/concurrent/TaskTest.scala#L222) fails reliably for me with errors like `java.lang.AssertionError: cedbfa !== ecdfba`.

However, it is testing that 6 independent threads execute in a certain order simply because they sleep for different times. It should just test that they all produced the result in some order, so that `sb.toString` is a permutation of `"abcdef"`.

Now, clearly this does not work on my machine, and it works on Travis for whatever reason (threads are non-deterministic, but maybe of their scheduler less so), but why would anyone expect this to succeed _reliably_ without proper synchronization?
"Don't ever do anything like this" seems rather basic, so what am I missing?

I already double-checked that `Nondeterminism[Future].chooseAny` runs tasks in parallel as expected (at least comments say so).

--- a/tests/src/test/scala/scalaz/concurrent/TaskTest.scala
+++ b/tests/src/test/scala/scalaz/concurrent/TaskTest.scala
@@ -6,7 +6,7 @@ import scalaz.scalacheck.ScalazArbitrary._
 import scalaz.std.AllInstances._
 import org.scalacheck.Prop._
 
-import java.util.concurrent.{Executors, TimeoutException}
+import java.util.concurrent.{Executors, TimeoutException, TimeUnit}
 import java.util.concurrent.atomic._
 import org.scalacheck.Prop.forAll
 
@@ -217,17 +217,20 @@ object TaskTest extends SpecLite {
     "nmap6 must run Tasks in parallel" in {
       import Thread._
       import java.{util => ju}
+      import ju.concurrent.CyclicBarrier
 
+      //Ensure at least 6 different threads are available.
       implicit val es6 =
         Executors.newFixedThreadPool(6)
+      val barrier = new CyclicBarrier(6);
 
       val seenThreadNames = scala.collection.JavaConversions.asScalaSet(ju.Collections.synchronizedSet(new ju.HashSet[String]()))
       val t =
-        for (i <- 0 to 5)
-          yield fork {
+        for (i <- 0 to 5) yield fork {
           seenThreadNames += currentThread().getName()
-          //Make it harder for es6 to reuse threads.
-          sleep(100)
+          //Prevent the execution scheduler from reusing threads. This will only
+          //proceed after all 6 threads reached this point.
+          barrier.await(1, TimeUnit.SECONDS)
           now(('a' + i).toChar)
         }
 

